crates/common/consensus/lean/src/state.rs

```diff
@@ -48,126 +59,343 @@ impl LeanState {
             })?;
 
-        // Take left side of the list (if any)
-        for (i, justification_bit) in self
-            .justifications_roots_validators
-            .iter()
-            .take(index * VALIDATOR_REGISTRY_LIMIT as usize)
-            .enumerate()
-        {
-            new_justifications_roots_validators
-                .set(i, justification_bit)
-                .map_err(|err| anyhow!("Failed to set new justification bit: {err:?}"))?;
+        // genesis block is always justified
+        let is_justified = self.latest_block_header.slot == 0;
+
+        self.justified_slots
+            .push(is_justified)
+            .map_err(|err| anyhow!("Failed to add to justified_slots: {err:?}"))?;
+
+        // if there were empty slots, push zero hash for those ancestors
+        let num_empty_slots = block.slot - self.latest_block_header.slot - 1;
+        for _ in 0..num_empty_slots {
+            self.historical_block_hashes
+                .push(B256::ZERO)
+                .map_err(|err| anyhow!("Failed to prefill historical_block_hashes: {err:?}"))?;
+
+            self.justified_slots
+                .push(false)
+                .map_err(|err| anyhow!("Failed to prefill justified_slots: {err:?}"))?;
         }
 
-        // Take right side of the list (if any)
-        for (i, justification_bit) in self
-            .justifications_roots_validators
-            .iter()
-            .skip((index + 1) * VALIDATOR_REGISTRY_LIMIT as usize)
-            .enumerate()
-        {
-            new_justifications_roots_validators
-                .set(
-                    index * VALIDATOR_REGISTRY_LIMIT as usize + i,
-                    justification_bit,
-                )
-                .map_err(|err| anyhow!("Failed to set new justification bit: {err:?}"))?;
+        // Cache current block as the new latest block
+        self.latest_block_header = BlockHeader {
+            slot: block.slot,
+            proposer_index: block.proposer_index,
+            parent_root: block.parent_root,
+            // Overwritten in the next process_slot call
+            state_root: B256::ZERO,
+            body_root: block.body.tree_hash_root(),
+        };
+
+        Ok(())
+    }
+
+    fn process_operations(&mut self, body: &BlockBody) -> anyhow::Result<()> {
+        // Process attestations
+        self.process_attestations(&body.attestations)?;
+        Ok(())
+    }
+
+    pub fn process_attestations(
+        &mut self,
+        attestations: &VariableList<Arc<SignedVote>, U4096>,
+    ) -> anyhow::Result<()> {
+        let mut justifications_map = self.get_justifications()?;
+
+        for signed_vote in attestations {
+            let vote = &signed_vote.data;
+            // Ignore votes whose source is not already justified,
+            // or whose target is not in the history, or whose target is not a
+            // valid justifiable slot
+            if !self.justified_slots[vote.source.slot as usize] {
```

Comment:
shouldn't we use get() instead of raw parses for bounds checks? and to avoid panics

Reply:
Updated in d7740d9